include/operations/blas_trees.hpp

```diff
@@ -0,0 +1,201 @@
+/***************************************************************************
+*  @license
+*  Copyright (C) 2016 Codeplay Software Limited
+*  Licensed under the Apache License, Version 2.0 (the "License");
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+*  For your convenience, a copy of the License has been included in this
+*  repository.
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+*
+*  SYCL-BLAS: BLAS implementation using SYCL
+*
+*  @filename blas_trees_expr.hpp
+*
+**************************************************************************/
+
+#ifndef BLAS_TREE_EXPR_HPP
+#define BLAS_TREE_EXPR_HPP
+
+#include <operations/blas_constants.hpp>
+#include <operations/blas_operators.hpp>
+
+namespace blas {
+
+/** Join.
+ * @brief Joins both sides of the expression in the single kernel.
+ */
+template <class LHS, class RHS>
+struct JoinExpr {
+  using value_type = typename RHS::value_type;
+
+  LHS l;
+  RHS r;
+
+  JoinExpr(LHS &_l, RHS _r) : l(_l), r(_r){};
+
+  // PROBLEM: Only the RHS size is considered. If LHS size is different??
+  // If it is smaller, eval function will crash
+  size_t getSize() const { return r.getSize(); }
+};
+
+/** Assign.
+ */
+template <class LHS, class RHS>
+struct AssignExpr {
+  using value_type = typename RHS::value_type;
+
+  LHS l;
+  RHS r;
+
+  AssignExpr(LHS &_l, RHS _r) : l(_l), r(_r){};
+
+  // PROBLEM: Only the RHS size is considered. If LHS size is different??
+  // If it is smaller, eval function will crash
+  size_t getSize() const { return r.getSize(); }
+};
+
+/*! DoubleAssign.
+ */
+template <class LHS1, class LHS2, class RHS1, class RHS2>
+struct DoubleAssignExpr {
+  LHS1 l1;
+  LHS2 l2;
+  RHS1 r1;
+  RHS2 r2;
+
+ public:
+  using value_type = typename RHS1::value_type;
+
+  DoubleAssignExpr(LHS1 &_l1, LHS2 &_l2, RHS1 _r1, RHS2 _r2)
+      : l1(_l1), l2(_l2), r1(_r1), r2(_r2){};
+
+  // PROBLEM: Only the RHS size is considered. If LHS size is different??
+  // If it is smaller, eval function will crash
+  size_t getSize() const { return r2.getSize(); }
+};
```

Comment:
I will leave it for further consideration, at the moment it should not be hard to change the code to achieve that functionality, so it is not critical.